### PR TITLE
Allow specifying a key for encrypted resource packs

### PR DESCRIPTION
**Pull request by BrandPVP**
## Introduction
To specify a resource encryption key, put the key in the [resource].key file alongside the resource pack. Example: vanilla.zip.key

* Allows the use of raw (non-printable) encryption keys
* Less easy to lose due to a configuration error
* Keys are associated with a pack name, rather than an entry in a config (e.g. if you change the order of packs, you don't need to remember to also move the encryption keys)

### Relevant issues
* #5258

### API changes
Added `\pocketmine\resourcepacks\ResourcePack->getEncryptionKey()`
It returns empty string if the pack has no key.

### CODE REVIEW

File: src/resourcepacks/ResourcePack.php [hunk 1]
```diff
@@ -45,6 +45,12 @@ public function getPackSize() : int;
 	 */
 	public function getPackVersion() : string;
 
+	/**
+	 * Returns the key with which the pack was encrypted.
+	 * @return string empty string if the pack is not encrypted
+	 */
+	public function getEncryptionKey() : string;
```
Comment (SOF3) [inline src/resourcepacks/ResourcePack.php]: This is an API breaking change.

**Comment by Lunarelly** [comment]
why just dont specify it in resourcepacks.yml?
like:
name: rp.zip
encrypted: true
key: yourkey

**Comment by dktapps** [comment]
Read the linked issue. It's discussed there.

File: src/network/mcpe/handler/ResourcePacksPacketHandler.php [hunk 2]
```diff
@@ -67,7 +67,17 @@ public function __construct(
...
-			return new ResourcePackInfoEntry($pack->getPackId(), $pack->getPackVersion(), $pack->getPackSize(), "", "", "", false);
+			$encryptionKey = $pack->getEncryptionKey();
+
+			return new ResourcePackInfoEntry(
+				$pack->getPackId(),
+				$pack->getPackVersion(),
+				$pack->getPackSize(),
+				$encryptionKey,
+				"",
+				$encryptionKey === "" ? "" : $pack->getPackId(),
```
Comment (dktapps) [inline src/network/mcpe/handler/ResourcePacksPacketHandler.php]: why is this necessary?

File: src/resourcepacks/ResourcePack.php (hunk 1 above)
Comment (dktapps) [reply src/resourcepacks/ResourcePack.php]: A better solution might be to add an EncryptedResourcePack interface that extends this one, or perhaps just keep the keys in ResourcePackManager directly.

File: src/resourcepacks/ResourcePackManager.php [hunk 3]
```diff
@@ -93,13 +95,19 @@ public function __construct(string $path, \Logger $logger){
 					throw new ResourcePackException("Directory resource packs are unsupported");
 				}
 
+				$encryptionKey = "";
```
Comment (dktapps) [inline src/resourcepacks/ResourcePackManager.php]: I think `null` would make more sense.

**Review by dktapps — CHANGES_REQUESTED** [review]
This can't be targeted at a stable release since it contains BC breaks.

Branch targets:
- `stable`: Bug fixes only
- `next-minor`: New features, refactors, changes that don't break the API
- `next-major`: Disruptive changes including API breaks

File: src/resourcepacks/ResourcePackManager.php [hunk 4]
```diff
@@ -49,6 +51,9 @@ class ResourcePackManager{
 	/** @var ResourcePack[] */
 	private array $uuidList = [];
 
+	/** @var array<string,string> */
```
Comment (dktapps) [inline src/resourcepacks/ResourcePackManager.php]:
```suggestion
	/** @phpstan-var array<string, string> */
```
Please also add an `@var string[]`.

They need to be maintained this way because some tools don't support this generic style of phpdoc.

File: src/resourcepacks/ResourcePackManager.php [hunk 5]
```diff
@@ -106,6 +111,11 @@ public function __construct(string $path, \Logger $logger){
 				if($newPack instanceof ResourcePack){
 					$this->resourcePacks[] = $newPack;
 					$this->uuidList[strtolower($newPack->getPackId())] = $newPack;
+
+					$keyPath = Path::join($this->path, $pack . ".key");
+					if(file_exists($keyPath)){
+						$this->encryptionKeys[strtolower($newPack->getPackId())] = Utils::assumeNotFalse(file_get_contents($keyPath));
```
Comment (dktapps) [inline src/resourcepacks/ResourcePackManager.php]: This will crash if the file permissions are incorrect. `ErrorToExceptionHandler::trapAndRemoveFalse()` should be used instead. Catch the `ErrorException` it throws and rethrow it as a `ResourcePackException`.

File: src/resourcepacks/ResourcePackManager.php [hunk 6]
```diff
@@ -153,4 +163,11 @@ public function getPackById(string $id) : ?ResourcePack{
 	public function getPackIdList() : array{
 		return array_keys($this->uuidList);
 	}
+
+	/**
+	 * Returns the key with which the pack was encrypted, or null if the pack has no key.
+	 */
+	public function getPackEncryptionKey(string $id) : ?string {
```
Comment (dktapps) [inline src/resourcepacks/ResourcePackManager.php]:
```suggestion
	public function getPackEncryptionKey(string $id) : ?string{
```

**Review by dktapps — COMMENTED** [review]
Looking better

File: src/network/mcpe/handler/ResourcePacksPacketHandler.php [hunk 7]
```diff
@@ -65,9 +65,19 @@ public function __construct(
...
-			return new ResourcePackInfoEntry($pack->getPackId(), $pack->getPackVersion(), $pack->getPackSize(), "", "", "", false);
+			$encryptionKey = $this->resourcePackManager->getPackEncryptionKey($pack->getPackId());
+
+			return new ResourcePackInfoEntry(
+				$pack->getPackId(),
+				$pack->getPackVersion(),
+				$pack->getPackSize(),
+				$encryptionKey ?? "",
+				"",
+				$encryptionKey === null ? "" : $pack->getPackId(), // ContentId is required for client-side validations when the pack is encrypted.
```
Comment (dktapps) [inline src/network/mcpe/handler/ResourcePacksPacketHandler.php]: Is there any reason not to just always specify this? I don't actually remember what it's for, but there's a similar Geyser PR for reference: https://github.com/GeyserMC/Geyser/pull/3263

**Review by dktapps — COMMENTED** [review]
Looks good, just one final thing

File: src/network/mcpe/handler/ResourcePacksPacketHandler.php (hunk 7 above)
Comment (BrandPVP) [reply src/network/mcpe/handler/ResourcePacksPacketHandler.php]: > Is there any reason not to just always specify this? I don't actually remember what it's for, but there's a similar Geyser PR for reference: [GeyserMC/Geyser#3263](https://github.com/GeyserMC/Geyser/pull/3263)

Idk, but this is completely useless if pack is not encrypted. Its set for marketplace packs

Comment (dktapps) [reply src/network/mcpe/handler/ResourcePacksPacketHandler.php]: I just don't think it makes much sense to special-case it, if it works for both encrypted and unencrypted.
